### PR TITLE
Add LinalgOnTensors xla_ops tests

### PR DESCRIPTION
**Pull request by asaadaldien**
- Move HLO -> HLO & Legalize workflow before LinalgOnTensors -> HLO
- Run canonicalization after HLO -> HLO.
- Create xla_ops tests.

### CODE REVIEW

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 1]
```diff
@@ -65,17 +65,18 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager) {
...
     passManager.addNestedPass<FuncOp>(createDispatchLinalgOnTensorsPass(
         clLinalgOnTensorsTileSizes, clLinalgOnTensorsEnableFusion));
   }
 
-  // Flatten structured control flow to our CFG.
-  passManager.addNestedPass<FuncOp>(mhlo::createLegalizeControlFlowPass());
-  passManager.addNestedPass<FuncOp>(createHLOPreprocessingPass());
-
   // Convert TOSA ops to Linalg-on-tensor ops.
   passManager.addNestedPass<FuncOp>(tosa::createTosaToLinalgOnTensors());
```
Comment (benvanik) [inline iree/compiler/Dialect/Flow/Transforms/Passes.cpp:81]: so we aren't playing whackamole here we need to fix this up properly asap - for now, can you at least move the passes that turn things into linalg above where we require things to be in linalg? (it was a nasty hack that it was added in that the top - it never belonged there)

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 2]
```diff
@@ -209,10 +210,8 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager) {
   // Note that as we are rematerializing things here it's critical we do not run
   // the canonicalizer/CSE between now and when we outline - otherwise it'll
   // undo all of our work!
-  if (!clEnableLinalgOnTensorsDispatch) {
-    passManager.addNestedPass<FuncOp>(
-        IREE::Flow::createRematerializeDispatchConstantsPass());
-  }
+  passManager.addNestedPass<FuncOp>(
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/Passes.cpp:213]: I would rather move this before line 175

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 3]
```diff
@@ -65,17 +65,18 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager) {
   //----------------------------------------------------------------------------
   passManager.addPass(createCanonicalizerPass());
 
+  // Flatten structured control flow to our CFG.
+  passManager.addNestedPass<FuncOp>(mhlo::createLegalizeControlFlowPass());
+  passManager.addNestedPass<FuncOp>(createHLOPreprocessingPass());
+
   // Frontload linalg-on-tensors transformations and dispatch region creation.
   if (clEnableLinalgOnTensorsDispatch) {
+    passManager.addNestedPass<FuncOp>(createCanonicalizerPass());
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/Passes.cpp:74]: I would rather move this if branch down to line 175 and do 

```
if (!clEnableLinalgOnTensorsDispatch) {
  ... // whatever exists from line 175 - 240 (or 230?) 
} else {
  passManager.addNestedPass<FuncOp>(createCanonicalizerPass());
  addHLOToLinalgOnTensorsPasses(passManager);
  passManager.addNestedPass<FuncOp>(createDispatchLinalgOnTensorsPass(
      clLinalgOnTensorsTileSizes, clLinalgOnTensorsEnableFusion));
}

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp (hunk 3 above)
Comment (asaadaldien) [reply iree/compiler/Dialect/Flow/Transforms/Passes.cpp:74]: Did that things (based on my understanding lines 175-240 for hlo/flow dispatch region creating) but not conversion to HAL is failing 
```
failed to legalize operation 'flow.dispatch.workgroups' that was explicitly marked illegal
```

Comment (asaadaldien) [reply iree/compiler/Dialect/Flow/Transforms/Passes.cpp:74]: The HAL translate pass isn't invoked this way. something is missing is baked in one of these passes prepares the executable for translation .

**Review by MaheshRavishankar — APPROVED** [review]
ok, some WIP things here. Let land it as is after tests pass. We can come back to the pass ordering shortly.

**Comment by asaadaldien** [comment]
FYI, 
This to track reordering flow passes: https://github.com/google/iree/issues/4708

**Pull request merged**